File: corpus/runs/pydantic-ai/exercise.py
"""Exercise one agent of the pinned Pydantic AI checkout so a real run can be joined to the declared graph.

The declared against exercised delta is the centre of this product and it had only ever been shown on the bundled
demonstration, which Orchescope also wrote. This runs third party code instead: the `bank_support` example of the
Pydantic AI repository, at the commit `corpus/corpus.yaml` pins, which declares one agent, one tool and one model.

No provider is called and no credential is needed. `TestModel` is the library's own offline model: it answers from the
tool schemas rather than from a network, and it calls every tool once, which is exactly the run this join needs.

Spans go wherever `OTEL_EXPORTER_OTLP_ENDPOINT` points, which is what `orchescope trace` sets before it runs a command.
Run it through that:

    orchescope --cwd corpus/.cache/pydantic-ai trace -- \
      corpus/.cache/venvs/pydantic-ai/bin/python corpus/runs/pydantic-ai/exercise.py

The environment is built by `scripts/corpus-runtime.mjs`.
"""

from __future__ import annotations

import os
import sqlite3
import sys
from pathlib import Path

# The example names an OpenAI model, and Pydantic AI resolves the provider while the module is being imported, which
# needs a key to be present before `TestModel` can replace anything. A placeholder is forced in rather than defaulted,
# so a real key in the environment cannot be picked up by a corpus run. `ALLOW_MODEL_REQUESTS` is turned off below,
# which makes any attempt to reach a provider raise instead of sending a request.
os.environ["OPENAI_API_KEY"] = "placeholder-no-request-is-made"

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

CHECKOUT = Path(sys.argv[1]).resolve() if len(sys.argv) > 1 else Path.cwd()
sys.path.insert(0, str(CHECKOUT / "examples"))

provider = TracerProvider()
provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter()))
trace.set_tracer_provider(provider)

from pydantic_ai import Agent, models  # noqa: E402
from pydantic_ai.models.test import TestModel  # noqa: E402

models.ALLOW_MODEL_REQUESTS = False

# Instrumentation is opt in per agent, and the example does not opt in, so it is turned on for every agent here.
Agent.instrument_all()

from pydantic_ai_examples import bank_support  # noqa: E402


def main() -> None:
    with sqlite3.connect(":memory:") as connection:
        cursor = connection.cursor()
        cursor.execute("CREATE TABLE customers(id, name, balance)")
        cursor.execute("INSERT INTO customers VALUES (123, 'John', 123.45)")
        connection.commit()

        # The example reads this module global from inside its `__main__` block, so it is set the same way here.
        bank_support.cur = cursor

        dependencies = bank_support.SupportDependencies(
            customer_id=123,
            db=bank_support.DatabaseConn(sqlite_conn=connection),
        )
        with bank_support.support_agent.override(model=TestModel()):
            result = bank_support.support_agent.run_sync(
                "What is my balance?", deps=dependencies
            )
        print(result.output)

    provider.force_flush()
    provider.shutdown()


if __name__ == "__main__":
    main()

File: corpus/runs/vercel-ai-chatbot/exercise.mjs
/**
 * Exercises the pinned Vercel AI SDK application so a JavaScript run can be joined to the declared graph.
 *
 * The join had only been shown on Python, and instrumentation is the half of it that differs most between languages:
 * a decoder that reads one dialect says nothing about the other. This drives the `ai-chatbot` checkout at the commit
 * `corpus/corpus.yaml` pins, using two pieces of that repository rather than reimplementations of them:
 *
 *   `lib/ai/models.mock.ts`     the repository's own offline model, which it uses for its own end to end tests
 *   `lib/ai/tools/get-weather.ts`  the repository's own tool, declared with the SDK's `tool()`
 *
 * No provider is called and no credential is needed. The tool is invoked with no arguments, which is the branch of
 * the repository's own code that answers without reaching a weather service, so a real third party tool executes and
 * nothing outside this machine is contacted.
 *
 * What this cannot do is run the application's own agents: `generateTitleFromUserMessage` and the chat route are
 * Next.js server actions that need a request, a session and a database. The agent in this run is therefore the
 * driver's own, and it arrives exercised and never declared. That is a result rather than a gap to hide, and it is
 * written up in docs/research/runtime-join-on-third-party-code.md.
 *
 * Spans go wherever OTEL_EXPORTER_OTLP_ENDPOINT points, which is what `orchescope trace` sets. Run it through that:
 *
 *   orchescope --cwd corpus/.cache/vercel-ai-chatbot trace -- \
 *     node corpus/runs/vercel-ai-chatbot/exercise.mjs <checkout> <node_modules>
 *
 * The environment is built by scripts/corpus/exercise.mjs.
 */

import { createRequire } from 'node:module';
import { pathToFileURL } from 'node:url';
import { join, resolve } from 'node:path';

const [, , checkoutArgument, modulesArgument] = process.argv;
const checkout = resolve(checkoutArgument ?? process.cwd());
const modules = resolve(modulesArgument ?? join(checkout, 'node_modules'));

/*
 * This driver lives in the Orchescope repository, so its own bare imports would resolve against Orchescope's
 * dependencies, which do not include the SDK. Resolving them from the cached tree instead makes the SDK the driver
 * drives the same instance the repository's own tool is built with. The checkout finds that tree by itself: it sits
 * one directory above every checkout, which is where Node looks next.
 */
const fromCache = createRequire(pathToFileURL(join(modules, 'noop.js')));
const cached = (specifier) => pathToFileURL(fromCache.resolve(specifier)).href;

const { OTLPTraceExporter } = await import(cached('@opentelemetry/exporter-trace-otlp-http'));
const { BatchSpanProcessor, NodeTracerProvider } = await import(
  cached('@opentelemetry/sdk-trace-node')
);
const { generateText, registerTelemetry, stepCountIs } = await import(cached('ai'));
const { OpenTelemetry } = await import(cached('@ai-sdk/otel'));

const provider = new NodeTracerProvider({
  spanProcessors: [new BatchSpanProcessor(new OTLPTraceExporter())],
});
provider.register();
// From version 7 the SDK emits nothing until a telemetry integration is registered, and this is that integration.
registerTelemetry(new OpenTelemetry());

/*
 * Node runs the checkout's TypeScript directly, and both of these modules are plain enough for type stripping: the
 * mock model imports only a type, and the tool imports the SDK and zod, which the checkout resolves upwards.
 */
const { chatModel } = await import(pathToFileURL(join(checkout, 'lib/ai/models.mock.ts')).href);
const { getWeather } = await import(
  pathToFileURL(join(checkout, 'lib/ai/tools/get-weather.ts')).href
);

/*
 * The repository's mock model always answers with text, because the end to end tests it was written for assert on
 * text. A run that never calls a tool cannot show a tool joining, so the first step is asked for by the driver and
 * every step after it is the repository's model answering. The tool that runs is the repository's own, and it is
 * called with no arguments: that is the branch of its own code which answers without contacting a weather service.
 */
let step = 0;
const model = {
  ...chatModel,
  doGenerate: (options) => {
    step += 1;
    if (step > 1) return chatModel.doGenerate(options);
    return Promise.resolve({
      content: [{ type: 'tool-call', toolCallId: 'corpus-1', toolName: 'getWeather', input: '{}' }],
      finishReason: 'tool-calls',
      usage: { inputTokens: 12, outputTokens: 8, totalTokens: 20 },
      warnings: [],
    });
  },
};

const result = await generateText({
  model,
  prompt: 'What is the weather?',
  tools: { getWeather },
  stopWhen: stepCountIs(3),
  experimental_telemetry: { functionId: 'corpus-exercise' },
});

// What the run did, so a corpus log says whether the tool was reached rather than only that the process exited.
process.stdout.write(`steps: ${result.steps.length}, tool calls: ${result.toolCalls.length}\n`);
await provider.shutdown();
